xds/internal/httpfilter/rbac/rbac.go

```diff
@@ -0,0 +1,162 @@
+		// validation failure if present." - A41
+		if policy.Condition != nil {
+			return nil, errors.New("rbac: Policy.condition is present")
+		}
+		if policy.CheckedCondition != nil {
+			return nil, errors.New("rbac: policy.CheckedCondition is present")
+		}
+
+		// "It is also a validation failure if Permission or Principal has a
+		// header matcher for a grpc- prefixed header name or :scheme." - A41
+		for _, principal := range policy.Principals {
+			if principal.GetHeader() != nil {
+				name := principal.GetHeader().GetName()
+				if name == ":scheme" || strings.HasPrefix(name, "grpc-") {
+					return nil, fmt.Errorf("rbac: principal header matcher for %v is :scheme or starts with grpc", name)
+				}
+			}
+		}
+		for _, permission := range policy.Permissions {
+			if permission.GetHeader() != nil {
+				name := permission.GetHeader().GetName()
+				if name == ":scheme" || strings.HasPrefix(name, "grpc-") {
+					return nil, fmt.Errorf("rbac: permission header matcher for %v is :scheme or starts with grpc", name)
+				}
+			}
+		}
+	}
+
+	return config{config: msg}, nil
+}
+
+func (builder) ParseFilterConfig(cfg proto.Message) (httpfilter.FilterConfig, error) {
+	return parseConfig(cfg)
+}
+
+func (builder) ParseFilterConfigOverride(override proto.Message) (httpfilter.FilterConfig, error) {
+	any, ok := override.(*anypb.Any)
+	if !ok {
+		return nil, fmt.Errorf("rbac: error parsing override config %v: unknown type %T", override, override)
+	}
+	msg := new(rpb.RBACPerRoute)
+	if err := ptypes.UnmarshalAny(any, msg); err != nil {
+		return nil, fmt.Errorf("rbac: error parsing override config %v: %v", override, err)
+	}
+	return parseConfig(msg.Rbac)
+}
+
+func (builder) IsTerminal() bool {
+	return false
+}
+
+var _ httpfilter.ServerInterceptorBuilder = builder{}
+
+// Optional interface builder implements in order to signify it works server side.
+func (builder) BuildServerInterceptor(cfg httpfilter.FilterConfig, override httpfilter.FilterConfig) (resolver.ServerInterceptor, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("rbac: nil config provided")
+	}
+
+	c, ok := cfg.(config)
+	if !ok {
+		return nil, fmt.Errorf("rbac: incorrect config type provided (%T): %v", cfg, cfg)
+	}
+
+	if override != nil {
+		// override completely replaces the listener configuration; but we
+		// still validate the listener config type.
+		c, ok = override.(config)
+		if !ok {
+			return nil, fmt.Errorf("rbac: incorrect override config type provided (%T): %v", override, override)
+		}
+	}
+
+	icfg := c.config
+	ce, err := rbac.NewChainEngine([]*v3rbacpb.RBAC{icfg.Rules})
+	if err != nil {
+		return nil, fmt.Errorf("error constructing matching engine: %v", err)
+	}
+	return &interceptor{chainEngine: ce}, nil
+}
+
+type interceptor struct {
+	chainEngine *rbac.ChainEngine
+}
+
+func (i *interceptor) AllowRPC(ctx context.Context) error {
+	// ":method can be hard-coded to POST if unavailable" - A41
```

Comment:
`:method` is already hard-coded to `POST`

https://github.com/grpc/grpc-go/blob/2608e38e6386be7400720fecf2ece176c4cbc1b2/internal/transport/http2_client.go#L485

And I think the gRFC is saying that even though the headers can include `:method`, the implementation can ignore that but hardcode `:method=POST`, but they also need to ensure the server denies requests for all other method types

This is what our server does:
https://github.com/grpc/grpc-go/blob/2608e38e6386be7400720fecf2ece176c4cbc1b2/internal/transport/http2_server.go#L469-L482

---

Also I would expect this to fail.
The transport will also add a `:method` header. Not sure what will happen if there are two `:method` headers.

Reply:
Yeah, that was actually my first PR I merged on this team :). Ok, I validated that server does that, but I thought transport didn't hardcode a method: POST header so I added it here.

Reply:
Yeah this is what I thought, the server transport validates that the method is post by pulling data off the hpack encoded field, but never writes to the metadata struct that it puts in the streams context. Should I just go ahead and append method to metadata in transport in operateHeaders()?

Reply:
Oh, I see what you are doing here.

Adding it to metadata in transport is fine.

Or maybe hardcode it in the `chainEngine.IsAuthorized`. So the engine will reject all non-`POST`?

Reply:
Since we know the transport rejects all non-POST, the requests that get to the interceptors can only be POST.
If the RBAC config says to reject POST, it can just reject everything. If the config is to reject non-POST, it's a no-op.

It's a hardcoded behavior because we know how the server transport works.

Adding `:method` in the transport would be a more general solution.

Reply:
Ok, I thought about this, I don't like the idea of having an extra check in matcher builders for this functionality. I think this makes more sense in a general solution way by adding this :method: POST to the transport. Will do that in  a separate PR as per my other comment.